pybossa/jobs.py

```diff
@@ -622,21 +622,99 @@ def send_mail(message_dict, mail_all=False):
             mail.send(message)
 
 
-def delete_bulk_tasks(data):
-    """Delete tasks in bulk from project."""
+def count_rows_to_delete(project_id):
     from sqlalchemy.sql import text
     from pybossa.core import db
-    import pybossa.cache.projects as cached_projects
-    from pybossa.cache.task_browse_helpers import get_task_filters
 
-    project_id = data['project_id']
-    project_name = data['project_name']
-    curr_user = data['curr_user']
-    coowners = data['coowners']
-    current_user_fullname = data['current_user_fullname']
-    force_reset = data['force_reset']
-    url = data['url']
-    params = {}
+    sql = text('''
+        SELECT COUNT(*) FROM task WHERE
+        project_id=:project_id;
+        ''')
+    params = dict(project_id=project_id)
+    total_tasks = db.session.execute(sql, params).scalar()
+
+    sql = text('''
+        SELECT COUNT(*) FROM task_run WHERE
+        project_id=:project_id;
+        ''')
+    params = dict(project_id=project_id)
+    total_taskrun = db.session.execute(sql, params).scalar()
+
+
+    sql = text('''
+        SELECT COUNT(*) FROM result WHERE
+        project_id=:project_id;
+        ''')
+    params = dict(project_id=project_id)
+    total_result = db.session.execute(sql, params).scalar()
+
+
+    sql = text('''
+        SELECT COUNT(*) FROM counter WHERE
+        project_id=:project_id;
+        ''')
+    params = dict(project_id=project_id)
+    total_counter = db.session.execute(sql, params).scalar()
+    return total_tasks, total_taskrun, total_result, total_counter
+
+
+def cleanup_task_records(project_id, limit):
+    """Cleanup records associated with task from all related tables."""
+
+    from sqlalchemy.sql import text
+    from pybossa.core import db
+
+    tables = ["counter", "result", "task_run", "task"]
+    task_ids = []
+    params = {"project_id": project_id, "limit": limit}
+    sql = text('''
+        SELECT id FROM task WHERE project_id=:project_id LIMIT :limit;
+        ''')
+    response = db.session.execute(sql, params).fetchall()
+    task_ids = [id[0] for id in response]
+    if not task_ids:
+        return
+
+    current_app.logger.info("Task ids staged for deletion: %s", task_ids)
+    task_ids_tuple = tuple(task_ids)
+    for table in tables:
+        sql = f"DELETE FROM {table} "
+        sql += "WHERE id IN :taskids;" if table == "task" else "WHERE task_id IN :taskids;"
+        db.session.execute(sql, {"taskids": task_ids_tuple})
+        db.session.commit()
+    current_app.logger.info("Task ids staged deleted from tables %s", tables)
+
+
+def delete_bulk_tasks_in_batches(project_id, force_reset):
+    """Delete bulk tasks in batches from project."""
+
+    batch_size = 100
```

Comment:
Consider making this a configurable parameter to adjust the batch size as needed, either defined at the top of the file or in settings_local.